plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,835 @@
+
+        # Will the simulation follow a uniform time step?
+        # This must be the case for certain stopping conditions and saving routines
+        self.is_uniform_time = isinstance(dt, u.Quantity) and isinstance(
+            dt.value, float
+        )
+
+        # Raise a ValueError if dt is required by stop condition or save routine but not specified
+        if (
+            stop_condition.require_uniform_dt or save_routine.require_uniform_dt
+        ) and not self.is_uniform_time:
+            raise ValueError(
+                "Please specify a uniform time step to use the simulation with this configuration!"
+            )
+
+        # Load and validate inputs
+        field_weightings = ["volume averaged", "nearest neighbor"]
+        if field_weighting in field_weightings:
+            self.field_weighting = field_weighting
+        else:
+            raise ValueError(
+                f"{field_weighting} is not a valid option for ",
+                "field_weighting. Valid choices are",
+                f"{field_weightings}",
+            )
+
+    def _tracked_particle_mask(self):
+        """
+        Calculates a boolean mask corresponding to particles that have not been stopped or removed.
+        """
+        return ~np.logical_or(np.isnan(self.x[:, 0]), np.isnan(self.v[:, 0]))
+
+    @property
+    def nparticles_tracked(self):
+        """Return the number of particles that don't have NaN position or velocity."""
+        return self._tracked_particle_mask().sum()
+
+    def run(
+        self,
+        stop_condition: AbstractStopCondition,
+        save_routine: Optional[AbstractSaveRoutine] = None,
+        dt=None,
+        field_weighting="volume averaged",
+    ):
+        r"""
+        Runs a particle-tracing simulation.
+        Timesteps are adaptively calculated based on the
+        local grid resolution of the particles and the electric and magnetic
+        fields they are experiencing.
+
+        Parameters
+        ----------
+        dt : `~astropy.units.Quantity`, optional
+            An explicitly set timestep in units convertible to seconds.
+            Setting this optional keyword overrules the adaptive time step
+            capability and forces the use of this timestep throughout. If a tuple
+            of timesteps is provided, the adaptive timestep will be clamped
+            between the first and second values.
+
+        field_weighting : str
+            String that selects the field weighting algorithm used to determine
+            what fields are felt by the particles. Options are:
+
+            * 'nearest neighbor': Particles are assigned the fields on
+                the grid vertex closest to them.
+            * 'volume averaged' : The fields experienced by a particle are a
+                volume-average of the eight grid points surrounding them.
+
+            The default is 'volume averaged'.
+
+        stop_condition : callable
```

Comment:
`function` -> `class`

Add a docstring entry for the save_routine

Maybe specify that this class is passed an instance of the particle tracer, and uses that to determine when to stop/save.